File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/find_the_duplicate_number/FindtheDuplicateNumber2.java
package main.java.com.jiang.leetcode100.find_the_duplicate_number;

import java.util.Scanner;

class Solution2 {
    public int findDuplicate(int[] nums) {
        int n = nums.length - 1;
        int max_bit = 31;
        while((n >> max_bit) == 0) max_bit--;
        int result = 0;
        for(int i = 0; i <= max_bit; i++) {
            int x = 0, y = 0;
            for(int j = 0; j <= n; j++) {
                if(((nums[j] >> i) & 1) == 1) x++;
                if(j > 0 && (((j >> i) & 1) == 1)) y++;
            }
            if(x > y) result |= (1 << i);
        }
        return result;
    }
}

public class FindtheDuplicateNumber2 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) nums[i] = Integer.parseInt(nums_str[i]);

        System.out.println(new Solution2().findDuplicate(nums));
    }
}

File: Leetcode100/leetcode100_python/FindtheDuplicateNumber2.py
from typing import List

class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        n = len(nums) - 1
        max_bit = 31
        while (n >> max_bit) == 0:
            max_bit -= 1
        
        result = 0
        for i in range(max_bit + 1):
            x, y = 0, 0
            for j in range(n + 1):
                if (nums[j] >> i) & 1:
                    x += 1
                if j > 0 and (j >> i) & 1:
                    y += 1
            if x > y:
                result |= 1 << i
        return result

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.findDuplicate(nums))